study1/week4/flood-fill.java

```diff
@@ -0,0 +1,49 @@
+//https://leetcode.com/problems/flood-fill/description/
+
+class Solution {
+    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
+        //time complexity is O(m*n=N) = O(N) (N is the total number of pixels in the image)
+        //space complexity is O(n), in the worst case, the queue may store all the pixels in the image
+
+        int originalColor  = image[sr][sc];
+        if (originalColor == color) {
+            return image;
+        }
+        
+        //direction
+        // int[] up = {-1, 0};
+        // int[] down = {1, 0};
+        // int[] r = {0, -1};
+        // int[] l = {0, 1};
+        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+
+        int n = image.length;
+        int m = image[0].length;
+
+        Queue<int[]> queue = new LinkedList<>();
+        queue.add(new int[]{sr, sc});
+
+        while (!queue.isEmpty()) {
+            int[] point = queue.poll();
+            int x = point[0];
+            int y = point[1];
+
+            image[x][y] = color;
+
+            for (int i = 0; i < directions.length; i++) {
+                int newX = x + directions[i][0];
+                int newY = y + directions[i][1];
+
+                if (isInRange(n, m, newX, newY) && image[newX][newY] == originalColor) {
```

Comment:
Iterating four directions in loop simplifies the code a bit, but it's not good for performance because potentially it's comparing the things that we don't need to compare. For instance for up (0, -1) you don't need to compare if row is bigger than 0 and smaller than maximum row because you're not changing the row at all. Also, you don't need to compare if the column is smaller than n because you are going up thus `n - 1` will always be smaller than the maximum column. This will be an unnecessary comparison. It's better in these case to unroll this loop and compare one by one.
```java
if (column > 0  && image[row][column - 1] == originalColor) {
   queue.add(new int[]{row, column - 1});
}
if (row> 0  && image[row - 1][column] == originalColor) {
   queue.add(new int[]{row - 1, column});
}
....
```